src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilationFailedResultFactory.cs

```diff
@@ -0,0 +1,159 @@
+                }
+
+                var compilationFailure = new CompilationFailure(
+                    sourceFilePath,
+                    sourceFileContent,
+                    compilationContent,
+                    group.Select(GetDiagnosticMessage),
+                    additionalMessage);
+
+                failures.Add(compilationFailure);
+            }
+
+            return new CompilationResult(failures);
+        }
+
+        private static string ReadContent(RazorCodeDocument codeDocument, string filePath)
+        {
+            RazorSourceDocument sourceDocument = null;
+            if (string.IsNullOrEmpty(filePath) || string.Equals(codeDocument.Source.FileName, filePath, StringComparison.Ordinal))
+            {
+                sourceDocument = codeDocument.Source;
+            }
+            else
+            {
+                sourceDocument = codeDocument.Imports.FirstOrDefault(f => string.Equals(f.FileName, filePath, StringComparison.Ordinal));
+            }
+
+            if (sourceDocument != null)
+            {
+                var contentChars = new char[sourceDocument.Length];
+                sourceDocument.CopyTo(0, contentChars, 0, sourceDocument.Length);
+                return new string(contentChars);
+            }
+
+            return string.Empty;
+        }
+
+        private static DiagnosticMessage GetDiagnosticMessage(Diagnostic diagnostic)
+        {
+            var mappedLineSpan = diagnostic.Location.GetMappedLineSpan();
+            return new DiagnosticMessage(
+                diagnostic.GetMessage(),
+                CSharpDiagnosticFormatter.Instance.Format(diagnostic),
+                mappedLineSpan.Path,
+                mappedLineSpan.StartLinePosition.Line + 1,
+                mappedLineSpan.StartLinePosition.Character + 1,
+                mappedLineSpan.EndLinePosition.Line + 1,
+                mappedLineSpan.EndLinePosition.Character + 1);
+        }
+
+        private static DiagnosticMessage CreateDiagnosticMessage(
+            RazorDiagnostic razorDiagnostic,
+            string filePath)
+        {
+            var sourceSpan = razorDiagnostic.Span;
+            var message = razorDiagnostic.GetMessage();
+            return new DiagnosticMessage(
+                message: message,
+                formattedMessage: razorDiagnostic.ToString(),
+                filePath: filePath,
+                startLine: sourceSpan.LineIndex + 1,
+                startColumn: sourceSpan.CharacterIndex,
+                endLine: sourceSpan.LineIndex + 1,
+                endColumn: sourceSpan.CharacterIndex + sourceSpan.Length);
+        }
+
+        private static string GetFilePath(RazorCodeDocument codeDocument, Diagnostic diagnostic)
+        {
+            if (diagnostic.Location == Location.None)
+            {
+                return codeDocument.Source.FileName;
+            }
+
+            return diagnostic.Location.GetMappedLineSpan().Path;
+        }
```

Comment:
Most of this code seems generic enough that it should live in `Microsoft.CodeAnalysis.Razor`